runtime/shared_common/shrinit.cpp

```diff
@@ -4923,4 +4931,201 @@ isFreeDiskSpaceLow(J9JavaVM *vm, U_64* maxsize)
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"-Dsun.java.launcher.pid="))
+		) {
+			keyLength += strlen(actualArgs->options[i].optionString);
+			extraLen += 1;
+		}
+	}
+	if (keyLength == 0) {
+		goto done;
+	}
+	keyLength += extraLen; /* space between options and terminating null char */
+
+	key = (char*)j9mem_allocate_memory(keyLength, J9MEM_CATEGORY_VM);
+	if (NULL == key) {
+		goto done;
+	}
+	memset(key, 0, keyLength);
+	for (i = 0; i < nOptions; ++i) {
+		char* option = actualArgs->options[i].optionString;
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"sun.java.launcher.pid"))
+		) {
+			if (firstOption) {
+				firstOption = false;
+				j9str_printf(PORTLIB, key, keyLength, "%s%s", key, option);
+			} else {
+				j9str_printf(PORTLIB, key, keyLength, "%s%s%s", key, " ", option);
+			}
+		}
+	}
+done:
+	return key;
+}
+
+
+/**
+ * This function fetches the startup hints from the shared cache and store it locally to vm->sharedClassConfig->localStartupHints.
+ * @param[in] currentThread  The current VM thread
+ *
+ */
+static void
+fetchStartupHintsFromSharedCache(J9VMThread* currentThread)
+{
+	J9JavaVM* vm = currentThread->javaVM;
+
+	if (NULL != vm->sharedClassConfig) {
+		char *key = NULL;
+		if (NULL != vm->sharedClassConfig->localStartupHints.key) {
+			/* key is not NULL, already fetched before, directly return */
+			return;
+		}
+		key = generateStartupHintsKey(vm);
+		if (NULL != key) {
+			J9SharedDataDescriptor dataDescriptor = {0};
+			if (0 < j9shr_findSharedData(currentThread, key, strlen(key), J9SHR_DATA_TYPE_STARTUP_HINTS, 0, &dataDescriptor, NULL)) {
+				Trc_SHR_Assert_True(J9SHR_DATA_TYPE_STARTUP_HINTS == dataDescriptor.type);
+				Trc_SHR_Assert_True(sizeof(J9SharedStartupHintsDataDescriptor) == dataDescriptor.length);
+				memcpy(&vm->sharedClassConfig->localStartupHints.hintsData, dataDescriptor.address, sizeof(J9SharedStartupHintsDataDescriptor));
+				vm->sharedClassConfig->localStartupHints.key = key;
+				Trc_SHR_INIT_fetchStartupHintsFromSharedCache_Hints_Found(currentThread, vm->sharedClassConfig->localStartupHints.hintsData.flags,
+						vm->sharedClassConfig->localStartupHints.hintsData.heapSize1, vm->sharedClassConfig->localStartupHints.hintsData.heapSize2);
+			} else {
+				Trc_SHR_INIT_fetchStartupHintsFromSharedCache_Hints_Not_Found(currentThread);
+			}
+		} else {
+			Trc_SHR_INIT_fetchStartupHintsFromSharedCache_Null_key(currentThread);
+		}
+	}
+}
+
+/**
+ * This function store the local startup hints vm->sharedClassConfig->localStartupHints to the shared cache
+ * @param[in] currentThread  The current VM thread
+ * @return  The location of the cached data or null
+ *
+ */
+const U_8*
+storeStartupHintsToSharedCache(J9VMThread* currentThread)
+{
+	J9JavaVM* vm = currentThread->javaVM;
+	char* key = vm->sharedClassConfig->localStartupHints.key;
+	const U_8 *ret = NULL;
+	if (NULL == key) {
```

Comment:
This function is always called, it should check there are hints to be written before generating the key.

Reply:
Fixed